app/src/commonMain/kotlin/io/github/cgr/app/VirtualMachine.kt

```diff
@@ -0,0 +1,31 @@
+package io.github.cgr.app
+
+class VirtualMachine(private val program: Program) {
+
+    var stack = ArrayDeque<Any?>()
+
+    private var pc = 0
+
+    fun run() {
+        while (pc < program.opcodes.size) {
+            val (opcode, args) = program.opcodes[pc++]
+            opcode.execute(args, this)
+        }
+    }
+
+    fun jump(address: Int) {
+        pc = address - 1 // -1 because pc will be incremented at the end of the loop
+    }
+
+    fun jumpRelative(offset: Int) {
+        pc += offset - 1 // -1 because pc will be incremented at the end of the loop
+    }
+
+    fun programCounter() = pc
```

Comment:
Is there a reason for doing this instead of
```kt
var programCounter = 0
    private set
```
? I'm not super familiar with this aspect of Kotlin.

Reply:
No. This is what PRs are for lol